### PR TITLE
fix: sonar issue that require use callback in setState

### PR DESCRIPTION
**Pull request by slin-sap**
Sonarqube rule explanation: https://sonarcloud.io/organizations/sap-1/rules?open=typescript%3AS6756&rule_key=typescript%3AS6756

Bugs reported in sonarqube: https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&types=BUG&id=SAP_open-ux-tools&open=AYr_XTAfzUdZB34-1Yfx&tab=why

### CODE REVIEW

File: packages/ui-components/src/components/UITreeDropdown/UITreeDropdown.tsx [hunk 1]
```diff
@@ -424,19 +424,21 @@ export class UITreeDropdown extends React.Component<UITreeDropdownProps, UITreeD
...
+                this.toggleMenu(false);
+                newState.isMenuOpen = false;
+            }
+            return newState;
         });
-
-        if (!this.state.isMenuOpen) {
-            this.toggleMenu(false);
-        }
     };
```
Comment (815are) [inline packages/ui-components/src/components/UITreeDropdown/UITreeDropdown.tsx:442]: I see following error:
image

Scenario:
1. Open storybook for tree drpdown;
2. Use tab key to navigate tree dropdown input - menu should not be opened;
3. Press char like `a` - it should trigger menu open;

I tried apply old version - I did not see such error

Comment (815are) [reply packages/ui-components/src/components/UITreeDropdown/UITreeDropdown.tsx]: @slin-sap
adjusted code with following - 0f92ca10ad71507f591599460c3cb919c755fb99
please correct if it provides issues

File: packages/ui-components/src/components/UITable/UITable.tsx [hunk 2]
```diff
@@ -676,26 +676,31 @@ export class UITable extends React.Component<UITableProps, UITableState> {
...
-            this.saveCell(false, newValue);
-        }
+        this.setState((prevState) => {
+            const editedCell = prevState.editedCell ?? this.activeElement;
+            if (editedCell) {
+                editedCell.newValue = newValue;
+
+                const column = editedCell?.column;
+                if (column && typeof column.validate === 'function') {
+                    this.validateCell(newValue);
```
Comment (815are) [inline packages/ui-components/src/components/UITable/UITable.tsx:686]: I also see same react warning with new code:
image

If I apply old code - there no such warning

Scenario:
1. open storybook
2. open editable table
3. change value of any cell to `hello` - it should trigger validation, which applied as demo in story

Problem - react warning appears

**Review by 815are — COMMENTED** [review]
Checked:
1. code changes - noticed `setState` calls within `setState` which gives warning mentioned in comment https://github.com/SAP/open-ux-tools/pull/1370/files#r1356443113;
Applied fix for `UITreeDropdown` from my side as familiar with logic.
2. Manually tested. UICombobox, UIDropdown, ToggleGroups, UITreeDropdown works as expected and did not find case with warning. UITable gives warning if validation message is applied( https://github.com/SAP/open-ux-tools/pull/1370/files#r1356443113)

**Comment by slin-sap** [comment]
> UITable gives warning if validation message is applied( https://github.com/SAP/open-ux-tools/pull/1370/files#r1356443113)

Thanks @815are. I will continue look the UITable warning. Did you do manual test on storybook?

**Comment by 815are** [comment]
> > UITable gives warning if validation message is applied( https://github.com/SAP/open-ux-tools/pull/1370/files#r1356443113)
> 
> Thanks @815are. I will continue look the UITable warning. Did you do manual test on storybook?

@slin-sap 
yes, I used storybook examples by launching `pnpm storybook`. And example under story https://github.com/SAP/open-ux-tools/blob/main/packages/ui-components/stories/UITable.story.tsx#L88

**Comment by slin-sap** [comment]
@815are 

I applied a fix for the validation scenario. I tested on storybook. It is able to displaying the validation UI on the cell without logging warning message in the console:

https://github.com/SAP/open-ux-tools/pull/1370/commits/a13dcfa625fc07d85eb83c58945660fb3ffb065c

I am still getting an error that crashes the entire app when I cancel edit. But can you take a look at my latest commit and see if the approach looks promising.

**Comment by 815are** [comment]
> @815are
> 
> I applied a fix [a13dcfa](https://github.com/SAP/open-ux-tools/commit/a13dcfa625fc07d85eb83c58945660fb3ffb065c) for the validation scenario. I tested on storybook. It is able to displaying the validation UI on the cell without logging warning message in the console.
> 
> I am still getting an error that crashes the entire app when I cancel edit. But can you take a look at my latest commit and see if the approach looks promising.

@slin-sap 
I restored original solution and applied correction on top - https://github.com/SAP/open-ux-tools/pull/1370/commits/0a530f938399941d987cddcfdaedd6f2537b643b
Please correct if this version provides issue

**Review by 815are — APPROVED** [review]
Initial review - https://github.com/SAP/open-ux-tools/pull/1370#pullrequestreview-1673602969
Follow up:
1. Fix for table applied;
2. Tested table component with edit/readonly modes and validation - did not find issue/regression. Also compared behaviour with current/main version Table - I did not notice difference in behavior;

**Review by 815are — APPROVED** [review]

**Review by devinea — COMMENTED** [review]
Thanks @slin-sap @815are  for the updates. 
@tom-sap Can you review UITable changes?

**Review by Jimmy-Joseph19 — COMMENTED** [review]
Thanks @slin-sap for the changes.
Changes looks compliant
Tested the changes using Storybook

**Comment by 815are** [comment]
> @815are
> 
> Thanks a lot for looking into this. The code changes look good to me. So for cancelEdit we are just setting editedCell to undefined and not actually reply on prevState, there is no need to apply my changes and fixed the error?

@slin-sap 
not fully sure I got question.
Do you mean `cancelEdit` call within `onTextInputChange/setState->saveCell`? or different scenario?

**Review by 815are — APPROVED** [review]

**Review by devinea — APPROVED** [review]
Thanks

**Review by Jimmy-Joseph19 — APPROVED** [review]
Thanks

**Pull request merged**